hail/src/main/scala/is/hail/expr/ir/ComputeUsesAndDefs.scala

```diff
@@ -2,95 +2,97 @@ package is.hail.expr.ir
+        case (child: BlockMatrixIR, _) => computeBlockMatrix(child, includeApplyIR)
       }
 
-    def computeMatrix(mir: MatrixIR): Unit = mir.children
+    def computeMatrix(mir: MatrixIR, includeApplyIR: Boolean): Unit = mir.children
       .iterator
       .zipWithIndex
       .foreach {
         case (child: IR, i) =>
           val b = NewBindings(mir, i).mapValues[BaseIR](_ => mir)
           if (!b.allEmpty && !uses.contains(mir))
-            uses.bind(mir, mutable.Set.empty[RefEquality[Ref]])
-          computeIR(child, b)
-        case (child: TableIR, _) => computeTable(child)
-        case (child: MatrixIR, _) => computeMatrix(child)
-        case (child: BlockMatrixIR, _) => computeBlockMatrix(child)
+            uses.bind(mir, mutable.Set.empty[RefEquality[BaseRef]])
+          computeIR(child, b, includeApplyIR)
+        case (child: TableIR, _) => computeTable(child, includeApplyIR)
+        case (child: MatrixIR, _) => computeMatrix(child, includeApplyIR)
+        case (child: BlockMatrixIR, _) => computeBlockMatrix(child, includeApplyIR)
       }
 
-    def computeBlockMatrix(bmir: BlockMatrixIR): Unit = bmir.children
+    def computeBlockMatrix(bmir: BlockMatrixIR, includeApplyIR: Boolean): Unit = bmir.children
       .iterator
       .zipWithIndex
       .foreach {
         case (child: IR, i) =>
           val b = NewBindings(bmir, i).mapValues[BaseIR](_ => bmir)
           if (!b.allEmpty && !uses.contains(bmir))
-            uses.bind(bmir, mutable.Set.empty[RefEquality[Ref]])
-          computeIR(child, b)
-        case (child: TableIR, _) => computeTable(child)
-        case (child: MatrixIR, _) => computeMatrix(child)
-        case (child: BlockMatrixIR, _) => computeBlockMatrix(child)
+            uses.bind(bmir, mutable.Set.empty[RefEquality[BaseRef]])
+          computeIR(child, b, includeApplyIR)
+        case (child: TableIR, _) => computeTable(child, includeApplyIR)
+        case (child: MatrixIR, _) => computeMatrix(child, includeApplyIR)
+        case (child: BlockMatrixIR, _) => computeBlockMatrix(child, includeApplyIR)
       }
 
-    def computeIR(ir: IR, env: BindingEnv[BaseIR]) {
+    def computeIR(ir: IR, env: BindingEnv[BaseIR], includeApplyIR: Boolean) {
       ir match {
-        case r@Ref(name, _) =>
-          env.eval.lookupOption(name) match {
+        case r: BaseRef =>
+          env.eval.lookupOption(r.name) match {
             case Some(decl) =>
               val re = RefEquality(r)
               uses.lookup(decl) += re
               defs.bind(re, decl)
             case None =>
               if (errorIfFreeVariables)
-                throw new RuntimeException(s"found variable with no definition: $name")
+                throw new RuntimeException(s"found variable with no definition: ${ r.name }")
           }
+        case _ =>
+      }
+      ir match {
+        case x: ApplyIR if includeApplyIR => computeIR(x.explicitNode, env, includeApplyIR)
```

Comment:
this seems wrong, though I understand why you've done it. The problem is really in inferring requiredness before ApplyIR has been inlined, right? We'd need to infer the loosest possible type there, without looking at the explicit node?

Reply:
Perhaps the right thing is to have ApplyIR's explicit node generate a function from argument TypeWithRequiredness to result TypeWithRequiredness.

Reply:
Is there a reason to avoid using x.explicitNode, specifically? I lifted this logic into the Requiredness class, but I'm still using x.body because there's not much I would do to implement this other than basically recreate what x.body does.